docs/specs/m2m.md

```diff
@@ -0,0 +1,1139 @@
+
+### Changes in data models
+
+Here are the schema of the changes:
+
+```sql
+CREATE TABLE _auth_resource (
+  id text PRIMARY KEY,
+  app_id text NOT NULL,
+  created_at timestamp without time zone NOT NULL,
+  updated_at timestamp without time zone NOT NULL,
+  uri text NOT NULL,
+  name text,
+  data jsonb
+);
+-- Each project has its own set of Resources. The URI must be unique within a project.
+CREATE UNIQUE INDEX _auth_resource_uri_unique ON _auth_resource USING btree (app_id, uri);
+-- Support typeahead search
+CREATE INDEX _auth_resource_uri_typeahead ON _auth_resource USING btree (app_id, uri text_pattern_ops);
+CREATE INDEX _auth_resource_name_typeahead ON _auth_resource USING btree (app_id, name text_pattern_ops);
+
+CREATE TABLE _auth_resource_scope (
+  id text PRIMARY KEY,
+  app_id text NOT NULL,
+  created_at timestamp without time zone NOT NULL,
+  updated_at timestamp without time zone NOT NULL,
+  resource_id text NOT NULL REFERENCES _auth_resource(id),
+  scope text NOT NULL,
+  description text,
+  data jsonb
+);
+-- Each Resource has its own set of Scopes. The scope must be unique within a Resource.
+CREATE UNIQUE INDEX _auth_resource_scope_unique ON _auth_resource_scope USING btree (app_id, resource_id, scope);
+-- Support typeahead search
+CREATE INDEX _auth_resource_scope_scope_typeahead ON _auth_resource_scope USING btree (app_id, resource_id, scope text_pattern_ops);
+
+CREATE TABLE _auth_client_resource (
+  id text PRIMARY KEY,
+  app_id text NOT NULL,
+  created_at timestamp without time zone NOT NULL,
+  updated_at timestamp without time zone NOT NULL,
+  -- Since client is not stored in the database, it is not a foreign key.
+  client_id text NOT NULL,
+  resource_id text NOT NULL REFERENCES _auth_resource(id),
+);
+-- Each Client can only associate with a Resource once.
+CREATE UNIQUE INDEX _auth_client_resource_unique ON _auth_client_resource USING btree (app_id, client_id, resource_id);
+
+CREATE TABLE _auth_client_resource_scope (
+  id text PRIMARY KEY,
+  app_id text NOT NULL,
+  created_at timestamp without time zone NOT NULL,
+  updated_at timestamp without time zone NOT NULL,
+  -- Since client is not stored in the database, it is not a foreign key.
+  client_id text NOT NULL,
+  resource_id text NOT NULL REFERENCES _auth_resource(id),
+  scope_id text NOT NULL REFERENCES _auth_resource_scope(id)
+);
+-- Each Client can only associate with a Resource scope once.
+CREATE UNIQUE INDEX _auth_client_resource_scope_unique ON _auth_client_resource USING btree (app_id, client_id, resource_id, scope_id);
+
+-- A sibling table of _auth_oauth_authorization, that takes resource_id into account.
+CREATE TABLE _auth_oauth_authorization_resource (
+  id text PRIMARY KEY,
+  app_id text NOT NULL,
+  client_id text NOT NULL,
+  user_id text NOT NULL REFERENCES _auth_user(id),
+  resource_id text NOT NULL REFERENCES _auth_resource(id),
+  scope_id text NOT NULL REFERENCES _auth_resource_scope(id)
+)
```

Comment:
@louischan-oursky 

I suggest to support a "{resource}/*" scope for each resource. For example, `https://onlinestore.myapp.com/*`.
It is because we probably don't want explicit assignment of scopes to each client. Say when you create an application, you want to allow it to access `https://onlinestore.myapp.com` with any permissions.
Then later, if you added new scopes to `https://onlinestore.myapp.com`, you probably won't remember to update the client you've created. Having a `*` scope in the client can automatically associate new scopes to the client.

Reply:
Added a commit to document our offline discussion on this suggestion.